### PR TITLE
SL-151/scroller

### PR DESCRIPTION
**Pull request by casserni**
adds scroll container that support custom theming

**NOTE**
If using a mac, you may not see these custom scroll bars and instead see the mac replaced one. To see these custom ones you may need to go to 
system preferences -> general -> Show Scrollbars-> always
This may also cause two sets of scroll bars to appear? Need to confirm

### CODE REVIEW

**Comment by marbemac** [comment]
Refactored, check it out - feels like a nicer pattern around exposing the inner functions like scrollToBottom, etc via a passed in innerRef (example in story). Also renamed component, figured we might want to rename `ListScroller` to be `ScrollList` for consistency?

File: src/theme/components.ts [hunk 1]
```diff
@@ -37,6 +37,10 @@ export const components: ISectionTheme = {
       border: '#/colors/transparent',
     },
 
+    scrollbar: {
```
Comment (P0lip) [inline src/theme/components.ts]: Could we add bg to https://github.com/stoplightio/ui-kit/blob/master/src/theme/components.ts#L23#L29?
I know it's consumed in a different component, but still, we could try to add it there? WDYT?
If that's not the case, I would consider renaming it to scrollContainer or something similar just to make it more clear.

File: src/ScrollBox.tsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            right: 10,
+            cursor: 'pointer',
+          }}
+        />
+      )}
+      renderThumbHorizontal={() => {
+        return <Box radius="full" height="6px" bg="scrollbar.bg" cursor="grab" />;
+      }}
+      renderTrackVertical={() => (
+        <div
```
Comment (P0lip) [inline src/ScrollBox.tsx]: Could we use Box here or do we want to avoid theming?

File: src/theme/components.ts (hunk 1 above)
Comment (casserni) [reply src/theme/components.ts]: I think that we might want to remove the scrollbars section in the top level of the theme. Allowing a user to change width and radius of the thumb is too much customizability. Really just allow users to change color

File: src/ScrollBox.tsx (hunk 2 above)
Comment (casserni) [reply src/ScrollBox.tsx]: Avoid theming on the track. Always hiding it makes for a cleaner ui. Would be a quick change to expose it though.

File: src/theme/components.ts (hunk 1 above)
Comment (P0lip) [reply src/theme/components.ts]: Agreed.

Comment (marbemac) [reply src/theme/components.ts]: Agreed - will leave it to you @casserni.

Comment (marbemac) [reply src/theme/components.ts]: Actually this gets back to the whole separating out "user customizable" portion of theme thing. We do want to maintain consistency with the look and feel of these scrollbars and https://github.com/stoplightio/ui-kit/blob/master/src/ListScroller.tsx#L69. Is there a better way than via theme variables that is not hardcoding duplicate values in both places?

**Comment by casserni** [comment]
Updated both ScrollBox and ScrollList to use our custom rendered toolbars. These both offset and hide the native toolbars so we won't need to worry about differences in browser behavior. Note that we should never use overflow scroll EVER and use ScrollBox in its place

File: src/__stories__/ScrollList.tsx [hunk 3]
```diff
@@ -4,30 +4,35 @@ import { array, number, NumberOptions, withKnobs } from '@storybook/addon-knobs'
...
+const rowRenderer = ({ value, key, style }: IScrollListItemProps) => (
   <Box key={key} as="div" {...style as object} borderBottom="2px solid black">
     {value}
   </Box>
 );
 
-storiesOf('ListScroller', module)
+storiesOf('ScrollList', module)
   .addDecorator(withKnobs)
+  // .addDecorator(storyFn => (
```
[the remark this answers is not in the document]
Comment (casserni) [reply src/__stories__/ScrollList.tsx]: Good catch

File: src/ScrollBox.tsx [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+              left: 2,
+            }}
+          />
+        );
+      }}
+      renderThumbHorizontal={({ style }: any) => {
+        return (
+          <Box
+            radius="full"
+            // @ts-ignore
```
[the remark this answers is not in the document]
Comment (casserni) [reply src/ScrollBox.tsx]: Don't left it in by accident

File: src/ScrollBox.tsx [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+          />
+        );
+      }}
+    >
+      {children}
+    </Scrollbars>
+  );
+};
+
+const useScrollTo = (elementId?: string) => {
```
[the remark this answers is not in the document]
Comment (marbemac) [reply src/ScrollBox.tsx]: Agreed.

File: src/ScrollList.tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      {({ height, width }) => (
+        <Box height={height} width={width} overflow="hidden" className={'box'}>
+          <List
+            className={className}
+            rowHeight={rowHeight}
+            rowCount={list.length}
+            rowRenderer={renderRow as ListRowRenderer}
+            noRowsRenderer={noRowsRenderer}
+            onScroll={(e: IOnScroll) => {
+              setClientHeight(e.clientHeight);
```
[the remark this answers is not in the document]
Comment (P0lip) [reply src/ScrollList.tsx]:
```js
const [{ clientHeight, scrollHeight, scrollTop }, setScrollEvent] = React.useState({});
```

something like that ^

[the remark this answers is not in the document]
Comment (casserni) [reply src/ScrollList.tsx]: Much cleaner thanks!

File: src/ScrollBox.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (casserni) [reply src/ScrollBox.tsx]: Moved this to the hooks folder and created some other utils

**Pull request merged**